### ISSUE TITLE
Conversion expressions fail to match correctly

### ISSUE THREAD

**Issue by GoogleCodeExporter** [ameniameni/moq#153]
```
What steps will reproduce the problem?
See sample code and test cases in attached file


What is the expected output? What do you see instead?
The setup should be matched, instead it is not when an implicit conversion 
has been genererated for an argument in the setup expression.


What version of the product are you using? On what operating system?
3.0.308.2


Please provide any additional information below.
An indication to a possible fix is included in the ConversionMatcher.cs 
file. This fix has not been verified using all the Moq tests, but it does 
fix the sample test cases provided.

Line 61 in MatcherFactory.cs is changed from:
    expression = ((UnaryExpression)expression).Operand;
to:
    return new ConversionMatcher((UnaryExpression)expression);
```

Original issue reported on code.google.com by `lukas.ma...@gmail.com` on 11 Mar 2009 at 1:23

Attachments:
- [ImplicitConversionProblem.cs](https://storage.googleapis.com/google-code-attachments/moq/issue-153/comment-0/ImplicitConversionProblem.cs)
- [ConversionMatcher.cs](https://storage.googleapis.com/google-code-attachments/moq/issue-153/comment-0/ConversionMatcher.cs)

**Comment by GoogleCodeExporter** [comment]
```
What's missing is the place where you hook the conversion matcher in the 
mockfactory. 
I'm not sure how it will work with the rest, as implicit conversion also 
happens when
you do: m.Do(It.IsAny<int>()) and you happen to change the parameter type in Do 
from
an int to a long, for example. In that case, your matcher will go against the
evaluation of the It.IsAny (0)...

I guess matchers together with convert should be forbidden?
```

Original comment by `kzu.net` on 12 Mar 2009 at 3:46

**Comment by GoogleCodeExporter** [comment]
```
This may be related, as a repro: http://www.pastie.org/415688
```

Original comment by `kzu.net` on 16 Mar 2009 at 1:53

**Comment by GoogleCodeExporter** [comment]
```
Has been fixed with #159.
```

Original comment by `kzu.net` on 16 Apr 2009 at 1:56
- Changed state: **Duplicate**
